File: src/main/java/com/github/unchama/targetedeffect/ops/TargetedEffectOps.kt
package com.github.unchama.targetedeffect.ops

import arrow.data.extensions.list.foldable.fold
import com.github.unchama.targetedeffect.TargetedEffect

operator fun <T> TargetedEffect<T>.plus(anotherEffect: TargetedEffect<T>): TargetedEffect<T> =
    with (TargetedEffect.monoid<T>()) { this@plus.combine(anotherEffect) }

fun <T> List<TargetedEffect<T>>.asSequentialEffect(): TargetedEffect<T> = fold(TargetedEffect.monoid())

File: src/main/java/com/github/unchama/targetedeffect/ops/TargetedEffectOps.scala
package com.github.unchama.targetedeffect.ops

import com.github.unchama.targetedeffect.TargetedEffect

import scala.collection.JavaConverters._

object TargetedEffectOps {
    implicit class TargetedEffectCombine[T](val effect: TargetedEffect[T]) {
        def +(anotherEffect: TargetedEffect[T]): TargetedEffect[T] =
            TargetedEffect.monoid[T]().combine(effect, anotherEffect)
    }

    implicit class TargetedEffectFold[T](val effects: List[TargetedEffect[T]]) {
        def asSequentialEffect(): TargetedEffect[T] = TargetedEffect.monoid[T]().combineAll(effects.asJava)
    }
}